neural_lam/datastore/mllam.py

```diff
@@ -0,0 +1,355 @@
+# Standard library
+import shutil
+import warnings
+from pathlib import Path
+from typing import List
+
+# Third-party
+import cartopy.crs as ccrs
+import mllam_data_prep as mdp
+import xarray as xr
+from loguru import logger
+from numpy import ndarray
+
+# Local
+from .base import BaseCartesianDatastore, CartesianGridShape
+
+
+class MLLAMDatastore(BaseCartesianDatastore):
+    """Datastore class for the MLLAM dataset."""
+
+    def __init__(self, config_path, n_boundary_points=30, reuse_existing=True):
+        """
+        Construct a new MLLAMDatastore from the configuration file at
+        `config_path`. A boundary mask is created with `n_boundary_points`
+        boundary points. If `reuse_existing` is True, the dataset is loaded
+        from a zarr file if it exists (unless the config has been modified
+        since the zarr was created), otherwise it is created from the
+        configuration file.
+
+        Parameters
+        ----------
+        config_path : str
+            The path to the configuration file, this will be fed to the
+            `mllam_data_prep.Config.from_yaml_file` method to then call
+            `mllam_data_prep.create_dataset` to create the dataset.
+        n_boundary_points : int
+            The number of boundary points to use in the boundary mask.
+        reuse_existing : bool
+            Whether to reuse an existing dataset zarr file if it exists and its
+            creation date is newer than the configuration file.
+
+        """
+        self._config_path = Path(config_path)
+        self._root_path = self._config_path.parent
+        self._config = mdp.Config.from_yaml_file(self._config_path)
+        fp_ds = self._root_path / self._config_path.name.replace(
+            ".yaml", ".zarr"
+        )
+
+        self._ds = None
+        if reuse_existing and fp_ds.exists():
+            # check that the zarr directory is newer than the config file
+            if fp_ds.stat().st_mtime > self._config_path.stat().st_mtime:
+                self._ds = xr.open_zarr(fp_ds, consolidated=True)
+            else:
+                logger.warning(
+                    "config file has been modified since zarr was created. "
+                    "recreating dataset."
+                )
+                shutil.rmtree(fp_ds)
+
+        if self._ds is None:
+            self._ds = mdp.create_dataset(config=self._config)
+            if reuse_existing:
+                self._ds.to_zarr(fp_ds)
```

Comment:
Undocumented effect of `reuse_existing`. Whether you reuse an existing zarr is an orthogonal option to whether you save a newly created one. I propose to create a separate argument for this (e.g. `save_to_disk`)  or just always defaulting to saving to disk.

Reply:
I am good with always defaulting to save to disk. We can add a config argument later than makes that optional

Reply:
removed `if reuse_existing` and took a note to add this as a seperate issue later